### PR TITLE
Feature/user session

### PR DESCRIPTION
**Pull request by starsinmypockets**
Use `USER_ACCOUNTS_ENABLED` in `.env` to enable account management routes:

* `/login`
* `/signup`
* `/logout`
* `/profile`

Internally, we use `req.session.ckan_user` to store logged user's API Key and profile info.

Logout deletes user from express session.

Implementer is responsible for exposing `user_login` method on CKAN Classic backend (see ckanext-nationalgrid for example)

Tests could be improved -- need to find a way to mock sessions to check creation / destruction of user on session (open to suggestions here!)

### CODE REVIEW

**Comment by starsinmypockets** [comment]
@anuveyatsu I'll see what I can do about test coverage. In the meantime your review will be greatly appreciated!

File: routes/user.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
+module.exports = function(app) {
+  const config = require('../config')
```
Comment (anuveyatsu) [inline routes/user.js]: Normally, I'd put all `require` in the top of the module.

File: routes/user.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  const { url, resolve } = require("url")
+  const fetch = require("node-fetch")
+  const proxy = require("express-http-proxy")
+  const { check, validationResult } = require("express-validator")
+  const { sanitizeBody } = require("express-validator/filter")
+
+  //  don't enable routes unless user accounts enabled in config
+  if (config.get("USER_ACCOUNTS_ENABLED")) {
+    const doAPILogin = async body => {
```
Comment (anuveyatsu) [inline routes/user.js]: I think this should go to `model` layer and be part of `/lib/user.js`, e.g., see how `dms` routes uses model `/lib/dms.js`.

File: routes/user.js [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        const APIResponse = await response.json()
+
+        return APIResponse && APIResponse.result
+      } catch (e) {
+        console.error("Error getting logged in user", e)
+        return false
+      }
+    }
+
+    const doAPIUserCreate = async body => {
```
Comment (anuveyatsu) [inline routes/user.js]: Also, in the model layer.

File: routes/user.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        }
+      }
+    )
+
+    app.get("/profile", (req, res) => {
+      if (!req.session.ckan_user) res.redirect("/")
+      res.render("user/profile.html", { user: req.session.ckan_user })
+    })
+
+    app.get("/login-failure", (req, res) => {
```
Comment (anuveyatsu) [inline routes/user.js]: I would prefer not having additional route for this but just showing a flash message on `/login` page.

File: views/user/login-failure.html [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{% extends "base.html" %}
+{% block content %}
+<div class="container">
+  <p>Login Failed</p>
```
Comment (anuveyatsu) [inline views/user/login-failure.html]: Can we make it a flash message on login page?

File: views/user/logged-out.html [hunk 6]
```diff
@@ -0,0 +1,4 @@
+{% extends "base.html" %}
+{% block content %}
+<p>You have been logged out! Come back soon.</p>
```
Comment (anuveyatsu) [inline views/user/logged-out.html]: Looks like this template isn't used. A flash message on home page would be great.

File: views/user/login-success.html [hunk 7]
```diff
@@ -0,0 +1,4 @@
+{% extends "base.html" %}
+{% block content %}
+<p>Welcome, {{ user.name }}!</p>
```
Comment (anuveyatsu) [inline views/user/login-success.html]: Is this used or we just show profile page?

File: views/user/user-menu.html [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{% if ckan_user %}
+<ul id="user-menu">
+  <li class="user-profile-icon">
+    <a href="/profile">Hello, {{ ckan_user.name }}</a>
```
Comment (anuveyatsu) [inline views/user/user-menu.html]: Probably this is a partial that will be used in the `base.html` as part of navbar?

File: index.js [hunk 9]
```diff
@@ -73,6 +75,9 @@ module.exports.makeApp = function () {
     }
   })
 
+  // Users
+  userRoutes(app)
```
Comment (anuveyatsu) [inline index.js]: Can't we define it as controller below? E.g.:

// Controllers
  app.use([
    dmsRoutes(),
    userRoutes()
])

Comment (starsinmypockets) [reply index.js]: I think this one is a little different because we need to pass the `app` object in order to enable some middleware.

Comment (anuveyatsu) [reply index.js]: ok, got it

File: routes/user.js [hunk 10]
```diff
@@ -33,10 +33,10 @@ module.exports = function(app) {
           const loggedUser = await Model.login(req.body)
           if (loggedUser) {
             req.session.ckan_user = loggedUser
-            req.flash('info', 'Testing flash messages')
             res.redirect("/profile")
           } else {
-            res.redirect("/login-failure")
+            req.flash('info', 'Login failed.')
```
Comment (anuveyatsu) [inline routes/user.js]: That's fine for now. We might need to provide details of the failure in the future.

**Pull request merged**